### PR TITLE
Simplify url validation regex and allow pluses

### LINKED ISSUE #18292: Can not add URL to asset (network device) as Link
### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

10.0.17

### Bug description

Hello there
If I try to add the following url as "Link" to an asset (network device), GLPI says invalid URL: https://knowledgebase.lancom-systems.de/display/KB/Erstellen+einer+%27Exposed+Host%27-Konfiguration+auf+einem+LANCOM+Router

Thanks in advance,
Michael

### Relevant log output

```shell

```

### Page URL

/front/manuallink.form.php?itemtype=NetworkEquipment&items_id=1

### Steps To reproduce

Create network device (only name and vendor)
Open device
switch to Links tab
Try to add URL from above as manual link

### Your GLPI setup information

<details><summary>Informationen über das System, die Installation und die Konfiguration</summary><pre>GLPI 10.0.17 ( => /var/www/glpi)
Installation mode: TARBALL
Current language:de_DE

</pre></details><details><summary>Server</summary><pre> 
Operating system: Linux gloin 6.1.0-25-amd64 #​1 SMP PREEMPT_DYNAMIC Debian 6.1.106-3 (2024-08-26) x86_64
PHP 8.2.24 fpm-fcgi (Core, FFI, PDO, Phar, Reflection, SPL, SimpleXML, Zend OPcache, bz2, calendar, cgi-fcgi, ctype, curl, date,
	dom, exif, fileinfo, filter, ftp, gd, gettext, hash, iconv, intl, json, ldap, libxml, mbstring, mysqli, mysqlnd, openssl, pcre,
	pdo_mysql, posix, random, readline, session, shmop, sockets, sodium, standard, sysvmsg, sysvsem, sysvshm, tokenizer, xml,
	xmlreader, xmlwriter, xsl, zip, zlib)
Setup: max_execution_time="30" memory_limit="128M" post_max_size="8M" safe_mode="" session.save_handler="files"
	upload_max_filesize="2M" disable_functions="" 
Software: Apache/2.4.62 (Debian) (Apache/2.4.62 (Debian) Server at gloin.techops.nwev.intern Port 443
)
	Mozilla/5.0 (X11; Linux x86_64; rv:132.0) Gecko/20100101 Firefox/132.0
Server Software: Debian 12
	Server Version: 10.11.6-MariaDB-0+deb12u1
	Server SQL Mode: STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION
	Parameters: glpi@localhost:3306/glpi
	Host info: Localhost via UNIX socket
	
PHP version (8.2.24) is supported.
Sessions configuration is OK.
Allocated memory is sufficient.
mysqli extension is installed.
Following extensions are installed: dom, fileinfo, filter, libxml, json, simplexml, xmlreader, xmlwriter.
curl extension is installed.
gd extension is installed.
intl extension is installed.
zlib extension is installed.
The constant SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_NPUBBYTES is present.
Database engine version (10.11.6) is supported.
No files from previous GLPI version detected.
The log file has been created successfully.
Write access to /var/www/glpi/files/_cache has been validated.
Write access to /var/www/glpi/files/_cron has been validated.
Write access to /var/www/glpi/files has been validated.
Write access to /var/www/glpi/files/_dumps has been validated.
Write access to /var/www/glpi/files/_graphs has been validated.
Write access to /var/www/glpi/files/_lock has been validated.
Write access to /var/www/glpi/files/_pictures has been validated.
Write access to /var/www/glpi/files/_plugins has been validated.
Write access to /var/www/glpi/files/_rss has been validated.
Write access to /var/www/glpi/files/_sessions has been validated.
Write access to /var/www/glpi/files/_tmp has been validated.
Write access to /var/www/glpi/files/_uploads has been validated.

Web server root directory configuration seems safe.
PHP directive "session.cookie_secure" should be set to "on" when GLPI can be accessed on HTTPS protocol.
PHP directive "session.cookie_httponly" should be set to "on" to prevent client-side script to access cookie values.
OS and PHP are relying on 64 bits integers.
exif extension is installed.
ldap extension is installed.
openssl extension is installed.
Following extensions are installed: bz2, Phar, zip.
Zend OPcache extension is installed.
Following extensions are installed: ctype, iconv, mbstring, sodium.
Write access to /var/www/glpi/marketplace has been validated.
Timezones seems loaded in database.

</pre></details><details><summary>GLPI constants</summary><pre> 
GLPI_ROOT: "/var/www/glpi"
GLPI_CONFIG_DIR: "/var/www/glpi/config"
GLPI_VAR_DIR: "/var/www/glpi/files"
GLPI_MARKETPLACE_DIR: "/var/www/glpi/marketplace"
GLPI_USE_CSRF_CHECK: "1"
GLPI_CSRF_EXPIRES: "7200"
GLPI_CSRF_MAX_TOKENS: "100"
GLPI_USE_IDOR_CHECK: "1"
GLPI_IDOR_EXPIRES: "7200"
GLPI_ALLOW_IFRAME_IN_RICH_TEXT: false
GLPI_SERVERSIDE_URL_ALLOWLIST: ["/^(https?|feed):\\/\\/[^@:]+(\\/.*)?$/"]
GLPI_TELEMETRY_URI: "https://telemetry.glpi-project.org"
GLPI_INSTALL_MODE: "TARBALL"
GLPI_NETWORK_MAIL: "glpi@teclib.com"
GLPI_NETWORK_SERVICES: "https://services.glpi-network.com"
GLPI_MARKETPLACE_ALLOW_OVERRIDE: true
GLPI_MARKETPLACE_MANUAL_DOWNLOADS: true
GLPI_USER_AGENT_EXTRA_COMMENTS: ""
GLPI_DISABLE_ONLY_FULL_GROUP_BY_SQL_MODE: "1"
GLPI_AJAX_DASHBOARD: "1"
GLPI_CALDAV_IMPORT_STATE: 0
GLPI_DEMO_MODE: "0"
GLPI_CENTRAL_WARNINGS: "1"
GLPI_TEXT_MAXSIZE: "4000"
GLPI_DOC_DIR: "/var/www/glpi/files"
GLPI_CACHE_DIR: "/var/www/glpi/files/_cache"
GLPI_CRON_DIR: "/var/www/glpi/files/_cron"
GLPI_DUMP_DIR: "/var/www/glpi/files/_dumps"
GLPI_GRAPH_DIR: "/var/www/glpi/files/_graphs"
GLPI_LOCAL_I18N_DIR: "/var/www/glpi/files/_locales"
GLPI_LOCK_DIR: "/var/www/glpi/files/_lock"
GLPI_LOG_DIR: "/var/www/glpi/files/_log"
GLPI_PICTURE_DIR: "/var/www/glpi/files/_pictures"
GLPI_PLUGIN_DOC_DIR: "/var/www/glpi/files/_plugins"
GLPI_RSS_DIR: "/var/www/glpi/files/_rss"
GLPI_SESSION_DIR: "/var/www/glpi/files/_sessions"
GLPI_TMP_DIR: "/var/www/glpi/files/_tmp"
GLPI_UPLOAD_DIR: "/var/www/glpi/files/_uploads"
GLPI_INVENTORY_DIR: "/var/www/glpi/files/_inventories"
GLPI_NETWORK_REGISTRATION_API_URL: "https://services.glpi-network.com/api/registration/"
GLPI_MARKETPLACE_PLUGINS_API_URI: "https://services.glpi-network.com/api/marketplace/"
GLPI_I18N_DIR: "/var/www/glpi/locales"
GLPI_VERSION: "10.0.17"
GLPI_SCHEMA_VERSION: "10.0.17@bde16719fbd4112f59a9a7d34c66c959bce73434"
GLPI_MARKETPLACE_PRERELEASES: false
GLPI_MIN_PHP: "7.4.0"
GLPI_MAX_PHP: "8.4.0"
GLPI_YEAR: "2024"

</pre></details><details><summary>Libraries</summary><pre> 
htmlawed/htmlawed version 1.2.14 in (/var/www/glpi/vendor/htmlawed/htmlawed)
phpmailer/phpmailer version 6.8.0 in (/var/www/glpi/vendor/phpmailer/phpmailer/src)
simplepie/simplepie version 1.5.8 in (/var/www/glpi/vendor/simplepie/simplepie/library)
tecnickcom/tcpdf version 6.7.5 in (/var/www/glpi/vendor/tecnickcom/tcpdf)
michelf/php-markdown in (/var/www/glpi/vendor/michelf/php-markdown/Michelf)
true/punycode in (/var/www/glpi/vendor/true/punycode/src)
iamcal/lib_autolink in (/var/www/glpi/vendor/iamcal/lib_autolink)
sabre/dav in (/var/www/glpi/vendor/sabre/dav/lib/DAV)
sabre/http in (/var/www/glpi/vendor/sabre/http/lib)
sabre/uri in (/var/www/glpi/vendor/sabre/uri/lib)
sabre/vobject in (/var/www/glpi/vendor/sabre/vobject/lib)
laminas/laminas-i18n in (/var/www/glpi/vendor/laminas/laminas-i18n/src)
laminas/laminas-servicemanager in (/var/www/glpi/vendor/laminas/laminas-servicemanager/src)
monolog/monolog in (/var/www/glpi/vendor/monolog/monolog/src/Monolog)
sebastian/diff in (/var/www/glpi/vendor/sebastian/diff/src)
donatj/phpuseragentparser in (/var/www/glpi/vendor/donatj/phpuseragentparser/src/UserAgent)
elvanto/litemoji in (/var/www/glpi/vendor/elvanto/litemoji/src)
symfony/console in (/var/www/glpi/vendor/symfony/console)
scssphp/scssphp in (/var/www/glpi/vendor/scssphp/scssphp/src)
laminas/laminas-mail in (/var/www/glpi/vendor/laminas/laminas-mail/src/Protocol)
laminas/laminas-mime in (/var/www/glpi/vendor/laminas/laminas-mime/src)
rlanvin/php-rrule in (/var/www/glpi/vendor/rlanvin/php-rrule/src)
ramsey/uuid in (/var/www/glpi/vendor/ramsey/uuid/src)
psr/log in (/var/www/glpi/vendor/psr/log/Psr/Log)
psr/simple-cache in (/var/www/glpi/vendor/psr/simple-cache/src)
psr/cache in (/var/www/glpi/vendor/psr/cache/src)
league/csv in (/var/www/glpi/vendor/league/csv/src)
mexitek/phpcolors in (/var/www/glpi/vendor/mexitek/phpcolors/src/Mexitek/PHPColors)
guzzlehttp/guzzle in (/var/www/glpi/vendor/guzzlehttp/guzzle/src)
guzzlehttp/psr7 in (/var/www/glpi/vendor/guzzlehttp/psr7/src)
glpi-project/inventory_format in (/var/www/glpi/vendor/glpi-project/inventory_format/lib/php)
wapmorgan/unified-archive in (/var/www/glpi/vendor/wapmorgan/unified-archive/src)
paragonie/sodium_compat in (/var/www/glpi/vendor/paragonie/sodium_compat/src)
symfony/cache in (/var/www/glpi/vendor/symfony/cache)
html2text/html2text in (/var/www/glpi/vendor/html2text/html2text/src)
symfony/css-selector in (/var/www/glpi/vendor/symfony/css-selector)
symfony/dom-crawler in (/var/www/glpi/vendor/symfony/dom-crawler)
twig/twig in (/var/www/glpi/vendor/twig/twig/src)
twig/string-extra in (/var/www/glpi/vendor/twig/string-extra)
symfony/polyfill-ctype not found
symfony/polyfill-iconv not found
symfony/polyfill-mbstring not found
symfony/polyfill-php80 not found
symfony/polyfill-php81 not found
symfony/polyfill-php82 in (/var/www/glpi/vendor/symfony/polyfill-php82)
league/oauth2-client in (/var/www/glpi/vendor/league/oauth2-client/src/Provider)
league/oauth2-google in (/var/www/glpi/vendor/league/oauth2-google/src/Provider)
thenetworg/oauth2-azure in (/var/www/glpi/vendor/thenetworg/oauth2-azure/src/Provider)

</pre></details><details><summary>SQL replicas</summary><pre> 
Not active

</pre></details><details><summary>Notifications</summary><pre> 
Way of sending emails: PHP

</pre></details><details><summary>Plugins list</summary><pre> 
	actualtime           Name: ActualTime                     Version: 3.0.0      State: Not installed                           
		Install Method: Marketplace
	advancedplanning     Name: advancedplanning               Version: 1.1.0      State: Installed / not activated               
		Install Method: Marketplace
	reports              Name: Berichte                       Version: 1.16.0     State: Installed / not activated               
		Install Method: Marketplace
	costs                Name: Costs                          Version: 3.0.3      State: Installed / not activated               
		Install Method: Marketplace
	mreporting           Name: More Reporting                 Version: 1.8.6      State: Installed / not activated               
		Install Method: Marketplace
	pdf                  Name: PDF-Ausgabe                    Version: 3.0.0      State: Installed / not activated               
		Install Method: Marketplace
	manufacturersimports Name: Suppliers imports              Version: 3.0.8      State: Not installed                           
		Install Method: Marketplace
	tag                  Name: Tag Management                 Version: 2.12.0     State: Installed / not activated               
		Install Method: Marketplace
	treeview             Name: Treeview                       Version: 1.10.2     State: Installed / not activated               
		Install Method: Marketplace

</pre></details>

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by cconard96**
## Checklist before requesting a review

- [x] I have read the CONTRIBUTING document.
- [x] I have performed a self-review of my code.
- [x] I have added tests that prove my fix is effective or that my feature works.

## Description

Fixes #18292
Pluses in a URL are allowed to be present unencoded to denote a space.

### CODE REVIEW

File: src/Toolbox.php [hunk 1]
```diff
@@ -3539,7 +3539,7 @@ public static function isValidWebUrl($url): bool
        // It is still possible to have false positives, but a fireproof check would be too complex
        // (or would require usage of a dedicated lib).
         return (preg_match(
-            "/^(?:http[s]?:\/\/(?:[^\s`!(){};'\",<>«»“”‘’+]+|[^\s`!()\[\]{};:'\".,<>?«»“”‘’+]))$/iu",
```
Comment (trasher) [inline src/Toolbox.php:3542]: regexps are always quite hard to read, but would'nt this change allows too much (including characters that should not be present in an URL)?

**Review by cedric-anne — CHANGES_REQUESTED** [review]
I did not remember how I get this regex pattern, but I am pretty sure that validating this change is pretty complicated.

We could here use `parse_url` to split the URL into its different parts and validate each part separately. It would probably be easier to read and to maintain.

The host validation should probably be more strict than the path validation.

```php
$parts = parse_url(Sanitizer::unsanitize($url));

if (!isset($parts['scheme']) || preg_match('/^https?$/', $parts['scheme']) !== 1) {
    return false;
}

...
```

Another solution would be to replace the pattern with a pattern used, for instance, by the URL validation in symfony (it uses a MIT licence, we would probably just have to add a link to the source file in our PHPDoc).
https://github.com/symfony/symfony/blob/7b20ed0c0fb5ac2bf52f6b9632d6b4985df61d04/src/Symfony/Component/Validator/Constraints/UrlValidator.php#L24-L46
I would be pretty confident with this solution, as I guess that if there was an issue on this pattern, it would already have been fixed.

**Comment by trasher** [comment]
I'd agree to use an existing, well tested solution.

**Comment by cconard96** [comment]
Just FYI, "parse_url" isn't reliable if it isn't a valid url to begin with according to the PHP docs.

File: phpunit/functional/ToolboxTest.php [hunk 2]
```diff
@@ -1188,7 +1188,7 @@ public static function urlProvider()
             ['https://localhost?test=true&othertest=false', true],
             ['https://localhost/front/computer.php?is_deleted=0&as_map=0&criteria[0][link]=AND&criteria[0][field]=80&criteria[0][searchtype]=equals&criteria[0][value]=254&search=Search&itemtype=Computer', true],
             ['https://localhost?test=true&#38;othertest=false', true], /* sanitized URL, &#38; is & */
-            ['https://localhost?test=true&#39;othertest=false', false], /* any entity */
```
Comment (trasher) [inline phpunit/functional/ToolboxTest.php:1191]: wgy removing this one?

**Review by trasher — APPROVED** [review]

File: phpunit/functional/ToolboxTest.php (hunk 2 above)
Comment (cconard96) [reply phpunit/functional/ToolboxTest.php:1191]: Symfony's regex says it is valid and it looks correct to me too. No reason an apostrphe couldn't be here, it just means the "othertest=false" isn't a separate parameter.

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**